File: element-common/src/main/java/io/github/solusmods/eternalcore/element/api/Elements.java
package io.github.solusmods.eternalcore.element.api;

import io.github.solusmods.eternalcore.element.impl.ElementsStorage;
import lombok.NonNull;
import net.minecraft.network.chat.MutableComponent;
import net.minecraft.resources.ResourceLocation;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Collection;
import java.util.Map;
import java.util.function.BiConsumer;

public interface Elements {

    Map<ResourceLocation, ElementInstance> getElements();

    Collection<ElementInstance> getObtainedElements();

    /**
     * Updates a element instance and optionally synchronizes the change across the network.
     * <p>
     * @param updatedInstance The instance to update
     * @param sync If true, synchronizes the change to all clients/server
     */
    void updateElement(ElementInstance updatedInstance, boolean sync);

    void forEachElement(BiConsumer<ElementsStorage, ElementInstance> skillInstanceConsumer);

    void forgetElement(@NotNull ResourceLocation skillId, @Nullable MutableComponent component);

    default void forgetElement(@NotNull ResourceLocation skillId) {
        forgetElement(skillId, null);
    }

    default void forgetElement(@NonNull Element element, @Nullable MutableComponent component) {
        forgetElement(element.getRegistryName(), component);
    }

    default void forgetElement(@NonNull Element element) {
        forgetElement(element.getRegistryName());
    }

    default void forgetElement(@NonNull ElementInstance instance, @Nullable MutableComponent component) {
        forgetElement(instance.getElementId(), component);
    }

    default void forgetElement(@NonNull ElementInstance instance) {
        forgetElement(instance.getElementId());
    }


    default boolean addElement(@NotNull ResourceLocation elementId, boolean teleportToSpawn) {
        return addElement(elementId, teleportToSpawn, null);
    }

    default boolean addElement(@NotNull ResourceLocation elementId, boolean teleportToSpawn, @Nullable MutableComponent component) {
        Element element = ElementAPI.getElementRegistry().get(elementId);
        if (element == null) return false;
        return addElement(element.createDefaultInstance(), false, teleportToSpawn, component);
    }

    default boolean addElement(@NonNull Element element, boolean teleportToSpawn) {
        return addElement(element, teleportToSpawn, null);
    }

    default boolean addElement(@NonNull Element element, boolean teleportToSpawn, @Nullable MutableComponent component) {
        return addElement(element.createDefaultInstance(), false, teleportToSpawn, component);
    }

    default boolean addElement(ElementInstance instance, boolean breakthrough, boolean teleportToSpawn) {
        return addElement(instance, breakthrough, teleportToSpawn, null);
    }

    boolean addElement(ElementInstance instance, boolean breakthrough, boolean teleportToSpawn, @Nullable MutableComponent component);
    
    void markDirty();

    void sync();
}

File: element-common/src/main/kotlin/io/github/solusmods/eternalcore/element/api/Elements.kt
package io.github.solusmods.eternalcore.element.api

import io.github.solusmods.eternalcore.element.impl.ElementsStorage
import net.minecraft.network.chat.MutableComponent
import net.minecraft.resources.ResourceLocation

interface Elements {
    val elements: MutableMap<ResourceLocation, ElementInstance>
    val obtainedElements: MutableCollection<ElementInstance>

    /**
     * Updates an element instance and optionally synchronizes the change across the network.
     *
     * @param updatedInstance The instance to update
     * @param sync If true, synchronizes the change to all clients/server
     */
    fun updateElement(updatedInstance: ElementInstance, sync: Boolean)

    fun forEachElement(action: (ElementsStorage, ElementInstance) -> Unit)

    fun forgetElement(elementId: ResourceLocation, component: MutableComponent? = null)

    fun forgetElement(element: Element, component: MutableComponent? = null) {
        forgetElement(element.registryName!!, component)
    }

    fun forgetElement(instance: ElementInstance, component: MutableComponent? = null) {
        forgetElement(instance.elementId, component)
    }

    fun addElement(
        elementId: ResourceLocation,
        teleportToSpawn: Boolean,
        component: MutableComponent? = null
    ): Boolean {
        val element = ElementAPI.elementRegistry.get(elementId) ?: return false
        return addElement(element.createDefaultInstance(), breakthrough = false, teleportToSpawn, component)
    }

    fun addElement(
        element: Element,
        teleportToSpawn: Boolean,
        component: MutableComponent? = null
    ): Boolean = addElement(element.createDefaultInstance(), breakthrough = false, teleportToSpawn, component)

    fun addElement(
        instance: ElementInstance,
        breakthrough: Boolean,
        teleportToSpawn: Boolean
    ): Boolean = addElement(instance, breakthrough, teleportToSpawn, component = null)

    fun addElement(
        instance: ElementInstance,
        breakthrough: Boolean,
        teleportToSpawn: Boolean,
        component: MutableComponent? = null
    ): Boolean

    fun markDirty()
    fun sync()
}